types/crossfilter2.d.ts

```diff
@@ -0,0 +1,103 @@
+declare namespace CrossFilter2 {
+
+    export interface Selector<T> {
+        (value: T): any;
+    }
+
+    export interface CrossFilterStatic {
+        <T>(data: T[]): CrossFilter<T>;
+        size(): number;
+        version: string;
+        permute<T>(arry: T[], index: number[], deep: number): T[];
+        heap: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Heap<T>;
+        }
+        heapselect: {
+            <T>(array: T[], lo: number, hi: number, k: number): T[];
+            by<T>(value: Selector<T>): HeapSelect<T>;
+        }
+        bisect: {
+            <T>(array: T[], value: T, lo: number, hi: number): number;
+            by<T>(value: Selector<T>): Bisector<T>;
+        }
+        insertionsort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(vaule: Selector<T>): Sort<T>;
+        }
+        quicksort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Sort<T>;
+        }
+    }
+
+    export interface Bisection<T> {
+        (array: T[], value: T, lo: number, hi: number): number;
+    }
+
+    export interface Bisector<T> extends Bisection<T> {
+        left: Bisection<T>
+        right: Bisection<T>
+    }
+
+    export interface Heap<T> {
+        (array: T[], lo: number, hi: number): T[];
+        sort(array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface HeapSelect<T> {
+        (array: T[], lo: number, hi: number, k: number): T[];
+    }
+
+    export interface Sort<T> {
+        (array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface CrossFilter<T> {
+        add(records: T[]): CrossFilter<T>;
+        remove(predicate?: Function): void;
+        dimension<T>(value: Function, isArray?: boolean): Dimension<T>;
+        groupAll(): any; // return vale?
+        size(): number;
+        all(): Array<T>;
+        allFiltered(): Array<T>;
+        onChange(callback: Function): Function;
+        isElementFiltered(index: number, ignoreDimensions?: number): boolean;
+
+    }
+
+    export interface Dimension<T> {
+        filter(range: any): any; // returns Array?
+        filterExact(value: any): any; // value most likely either string or number? Returns Array?
+        filterRange(range: T[]): any; // returns Array?
+        filterFunction(funct: Function): any;
+        filterAll(): any; // returns Array?
+        top(k: number, offset?: number): T[];
+        bottom(k: number, offset?: number): T[];
+        group(groupValue: Function): Group<T>;
```

Comment:
The argument to group is a function with a bunch of arguments, most of them optional. Would be good to get that figured out. I don't remember off the top of my head, but I think it's in the documentation. Sorry not to be more helpful.

Reply:
If the `interface` function is the correct TypeScript then it won't be a problem.  Just need confirmation that is my thoughts are the correct way of going so I can implement it.

Reply:
I think you're right about interface functions. Sorry I can't give much more guidance than that. It is a bit beyond my level of experience with Typescript.